### DISCUSSION TITLE
WebContainer, file system APIs & sqlite

### DISCUSSION THREAD

**Discussion by seidtgeist** [rhashimoto/wa-sqlite#19] [General]
It seems like a bunch of stuff was released for Google I/O ;) Something that caught my attention:

WebContainer OSS repo (docs only): https://github.com/stackblitz/webcontainer-core
HN thread: https://news.ycombinator.com/item?id=27223012
Blog post: https://blog.stackblitz.com/posts/introducing-webcontainers/
Comment mentioning sqlite: https://news.ycombinator.com/item?id=27224683

I’m curious about WebContainers appearing out of the blue, what exact file system API is being used ([File System Access API?](https://web.dev/file-system-access/)) and whether threads are involved in their sqlite adaption. As always, I’m hoping a proper FS API and WASM threads become available cross-browser.

**Comment by phiresky** [comment]
It is the FS Access API since www.StackBlitz.com/local throws `TypeError: window.showDirectoryPicker is not a function`

**Comment by rhashimoto** [comment]
I got excited enough to look at the File System Access API when support recently showed up in Safari, but unfortunately it doesn't look good for storing live databases. Besides the fact that [Firefox is opposing it](https://github.com/mozilla/standards-positions/pull/545), the Chromium implementation appears to be crippling performance in the name of security. According to [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1168715&q=file%20system%20access&can=5), writing to a file copies the file, applies the write to the copy, scans the copy for malware, and renames the copy over the original. If you're updating a small area of a large file, e.g. like a database transaction, and your local filesystem doesn't do some sort of block deduplication, then that could be excruciatingly slow. So far the Google response seems to be, "works as intended".

**Reply by rhashimoto** [reply]
Chrome 99 is shipping the [Origin Private File System](https://chromestatus.com/feature/5702777582911488) which might have much better I/O performance. As implied by the name, it is designed to be accessed only by origin pages, so it can't (or at least shouldn't) be used to share files with native applications. This is already in Safari, and is getting [a better initial reception from Firefox](https://github.com/mozilla/standards-positions/issues/562).

**Reply by seidtgeist** [reply]
@rhashimoto Here’s another dusty old idea: Have a sync node on LAN, e.g. an iOS/Anroid app and use that for persistent storage. I’ve given up my belief in persistent storage long ago.

**Reply by rhashimoto** [reply]
Long ago? You just started this thread with hope in May! 😀 

> As always, I’m hoping a proper FS API and WASM threads become available cross-browser.

Or do you mean shared persistent storage?

**Reply by seidtgeist** [reply]
Persistent storage is the Dark Souls of Browser APIs 😅 
The native-app-as-sync-node idea is from a October 2016 @BerlinJS talk I held 🥲

> Or do you mean shared persistent storage?

Single-tenant persistent storage.

**Reply by rhashimoto** [reply]
> Chrome 99 is shipping the [Origin Private File System](https://chromestatus.com/feature/5702777582911488)

This was misreported because they postponed the target release from 99 to 102 but forgot to update their feature calendar. So it's only released by Safari right now, still at the origin trial stage in Chrome. Chrome 102 is [scheduled to release](https://chromiumdash.appspot.com/schedule) in late May. Perhaps there will also be [progress in Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1748667) by then. 🤞
